File: components/credentials/src/main/java/org/cloudfoundry/credhub/utils/PrivateKeyReader.java
package org.cloudfoundry.credhub.utils;

import java.io.IOException;
import java.io.StringReader;
import java.security.PrivateKey;
import java.security.PublicKey;

import org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
import org.bouncycastle.openssl.PEMKeyPair;
import org.bouncycastle.openssl.PEMParser;
import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;

final public class PrivateKeyReader {

  private PrivateKeyReader() {
    super();
  }

  public static PrivateKey getPrivateKey(final String privateKeyPem) throws IOException, UnsupportedFormatException {
    final PEMParser pemParser = new PEMParser(new StringReader(privateKeyPem));
    final Object parsed = pemParser.readObject();
    pemParser.close();
    if (!(parsed instanceof PEMKeyPair)) {
      throw new UnsupportedFormatException("format of private key is not supported.");
    }
    final PEMKeyPair pemKeyPair = (PEMKeyPair) parsed;
    final PrivateKeyInfo privateKeyInfo = pemKeyPair.getPrivateKeyInfo();
    return new JcaPEMKeyConverter().getPrivateKey(privateKeyInfo);
  }

  public static PublicKey getPublicKey(final String privateKeyPem) throws IOException, UnsupportedFormatException {
    final PEMParser pemParser = new PEMParser(new StringReader(privateKeyPem));
    final Object parsed = pemParser.readObject();
    pemParser.close();
    if (!(parsed instanceof PEMKeyPair)) {
      throw new UnsupportedFormatException("format of private key is not supported.");
    }
    final PEMKeyPair pemKeyPair = (PEMKeyPair) parsed;
    final SubjectPublicKeyInfo publicKeyInfo = pemKeyPair.getPublicKeyInfo();
    return new JcaPEMKeyConverter().getPublicKey(publicKeyInfo);
  }

  public static class UnsupportedFormatException extends Exception {
    private static final long serialVersionUID = -2669429797326574839L;

    public UnsupportedFormatException(final String msg) {
      super(msg);
    }
  }
}

File: components/credentials/src/main/kotlin/org/cloudfoundry/credhub/utils/PrivateKeyReader.kt
package org.cloudfoundry.credhub.utils

import org.bouncycastle.asn1.pkcs.PrivateKeyInfo
import org.bouncycastle.openssl.PEMException
import org.bouncycastle.openssl.PEMKeyPair
import org.bouncycastle.openssl.PEMParser
import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter
import org.cloudfoundry.credhub.exceptions.MalformedPrivateKeyException
import java.io.StringReader
import java.security.KeyFactory
import java.security.PrivateKey
import java.security.PublicKey
import java.security.interfaces.RSAPrivateCrtKey
import java.security.interfaces.RSAPrivateKey
import java.security.spec.RSAPublicKeySpec

class PrivateKeyReader private constructor() {
    companion object {
        @JvmStatic
        fun getPrivateKey(privateKeyPem: String): PrivateKey {
            val pemParser: PEMParser
            val parsed: Any?
            try {
                pemParser = PEMParser(StringReader(privateKeyPem))
                parsed = pemParser.readObject()
                pemParser.close()
            } catch (e: PEMException) {
                throw MalformedPrivateKeyException("Keys must be PEM-encoded PKCS#1 or unencrypted PKCS#8 keys.")
            }

            return when (parsed) {
                // PKCS1
                is PEMKeyPair -> JcaPEMKeyConverter().getPrivateKey(parsed.privateKeyInfo)
                // PKCS8
                is PrivateKeyInfo -> JcaPEMKeyConverter().getPrivateKey(parsed)
                else -> throw MalformedPrivateKeyException("Key file is not in PKCS#1 or unencrypted PKCS#8 format")
            } as? RSAPrivateKey ?: throw MalformedPrivateKeyException("Key file does not contain an RSA private key")
        }

        @JvmStatic
        fun getPublicKey(privateKeyPem: String): PublicKey {
            val privateKey = getPrivateKey(privateKeyPem) as RSAPrivateCrtKey
            val publicKeySpec = RSAPublicKeySpec(privateKey.modulus, privateKey.publicExponent)

            val keyFactory = KeyFactory.getInstance("RSA")
            return keyFactory.generatePublic(publicKeySpec)
        }
    }
}
